### PR TITLE
Support download of binary artifacts

### PR DESCRIPTION
**Pull request by anders-kiaer**
...e.g. from GitHub artifacts created by GitHub workflows on GitHub releases.

### CODE REVIEW

**Comment by jondequinor** [comment]
I like it, but I am not a fan of the deferral to bash, since it makes things a tad more oblique. 

```diff
diff --git a/komodo/build.py b/komodo/build.py
index 26efda9..fa35066 100644
--- a/komodo/build.py
+++ b/komodo/build.py
@@ -115,6 +115,24 @@ def rsync(pkg, ver, pkgpath, data, prefix, *args, **kwargs):
         ]
     )
 
+def install_binary(pkg, ver, pkgpath, data, prefix, *args, **kwargs):
+    print('Installing {} ({}) with install_binary'.format(pkg, ver))
+    dest_name = pkgpath
+    opts = kwargs.get("makeopts", "").split(" ")
+    for idx, opt in enumerate(opts):
+        if opt == "--save-as":
+            dest_name = opts[idx+1]
+            opts = opts[:idx] + opts[idx+2:]
+            break
+    shell(
+        [
+            "rsync -ad",
+            " ".join(opts),
+            "{}".format(pkgpath),
+            kwargs["fakeroot"] + prefix + "/bin/" + dest_name,
+        ]
+    )
+
 
 def pip_install(pkg, ver, pkgpath, data, prefix, dlprefix, pip="pip", *args, **kwargs):
     ver = strip_version(ver)
@@ -197,7 +215,7 @@ def make(
     def resolve(x):
         return x.replace('$(prefix)', prefix)
 
-    build = { 'rpm': rpm, 'cmake': cmake, 'sh': sh, 'pip': pip_install, 'rsync': rsync, 'noop': noop }
+    build = { 'rpm': rpm, 'cmake': cmake, 'sh': sh, 'pip': pip_install, 'rsync': rsync, 'noop': noop, 'install_binary': install_binary }
 
     for pkg, path in zip(pkgorder, pkgpaths):
         ver = pkgs[pkg]
diff --git a/komodo/data/download_binary_asset.sh b/komodo/data/download_binary_asset.sh
deleted file mode 100644
index e311619..0000000
--- a/komodo/data/download_binary_asset.sh
+++ /dev/null
@@ -1,28 +0,0 @@
-#!/bin/sh
-set -e
-
-while test $# -gt 0; do
-    case "$1" in
-        --prefix)
-            shift
-            export PREFIX=$1
-            ;;
-        --fakeroot)
-            shift
-            export FAKEROOT=$1
-            ;;
-        --url)
-            shift
-            export URL=$1
-            ;;
-        --save-as)
-            shift
-            export SAVE_AS=$1
-            ;;
-    esac
-    shift
-done
-
-FAKEPREFIX=${FAKEROOT}/${PREFIX}
-mkdir -p "${FAKEPREFIX}/bin/"
-wget -O "${FAKEPREFIX}/bin/${SAVE_AS}" $URL
diff --git a/tests/data/cli/nominal_repository.yml b/tests/data/cli/nominal_repository.yml
index d98b323..0f825e3 100644
--- a/tests/data/cli/nominal_repository.yml
+++ b/tests/data/cli/nominal_repository.yml
@@ -53,7 +53,7 @@ python:
 
 some_github_binary_artifact:
   0.0.1a1:
-    make: sh
-    makefile: download_binary_asset.sh
+    make: install_binary
     maintainer: alice
-    makeopts: --url https://github.com/equinor/webviz-config-editor/releases/download/v0.0.1-alpha.1/WebvizConfigEditor-linux-0.0.1-x86_64.AppImage --save-as some-github-binary-artifact
+    source: https://github.com/equinor/webviz-config-editor/releases/download/v0.0.1-alpha.1/WebvizConfigEditor-linux-0.0.1-x86_64.AppImage
+    makeopts: --save-as some-github-binary-artifact
```

**Comment by anders-kiaer** [comment]
Nice - I wasn't aware of how to not use bash in Komodo. I'll try to switch to Python as suggested.

**Comment by pinkwah** [comment]
Maybe we could change the syntax?
```yaml
some_github_binary_artifact:
  0.0.1a1:
    make: download
    url: https://github.com/equinor/webviz-config-editor/releases/download/v0.0.1-alpha.1/WebvizConfigEditor-linux-0.0.1-x86_64.AppImage
    destination: bin/some-github-binary-artifact
    maintainer: alice
```

I would also like to have a CRC check and maybe be able to set the executable bit.

File: komodo/build.py [hunk 1]
```diff
@@ -116,6 +120,50 @@ def rsync(pkg, ver, pkgpath, data, prefix, *args, **kwargs):
     )
 
 
+def download(pkg, ver, pkgpath, data, prefix, *args, **kwargs):
+    print(f"Installing {pkg} ({ver}) with download")
+
+    url = kwargs["url"]
+    if not url.startswith("https"):
+        raise ValueError(f"{url} does not use https:// protocol")
+
```
Comment (jondequinor) [inline komodo/build.py:130]: If `hash` is required then you probably need to 
```suggestion
if not "hash" in kwargs:
    raise ValueError("no hash provided")
```

Comment (jondequinor) [reply komodo/build.py:130]: This is almost exactly equal to the KeyError, so this is up to you.

**Review by jondequinor — APPROVED** [review]
Seems solid! One small comment and then 🚀

File: komodo/build.py [hunk 2]
```diff
@@ -197,13 +246,20 @@ def make(
...
         ver = pkgs[pkg]
         current = repo[pkg][ver]
         make = current['make']
         pkgpath = os.path.abspath(path)
+
+        download_keys = ["url", "destination", "hash"]
+        if any(key in current for key in download_keys) and make != "download":
+            raise ValueError(", ".join(download_keys) + " only valid with 'make: download'")
+        if not all(key in current for key in download_keys) and make == "download":
+            raise ValueError(", ".join(download_keys) + " all required with 'make: download'")
```
Comment (jondequinor) [inline komodo/build.py:261]: Good!

**Comment by anders-kiaer** [comment]
Thanks for review :slightly_smiling_face: - fixup commits squashed. Also added an extra check that all required `.yml` keys are present with `download`.

I don't have write access so someone with :godmode: needs to click merge. :checkered_flag:

**Comment by jondequinor** [comment]
I am able to build a local komodo distribution, so will do an actual test of this. I'll also explain how I did it.

**Comment by jondequinor** [comment]
@anders-kiaer:

```shell
# Create a release file we can use to create a smaller distro
cd komodo-releases
komodo-transpiler transpile --matrix-file releases/matrices/2022.04.00.yml --output-folder .

# Append a package to repository.yml that will download stuff
echo "
some_github_binary_artifact:
  0.0.1a1:
    make: download
    maintainer: alice
    url: https://github.com/equinor/webviz-config-editor/releases/download/v0.0.1-alpha.1/WebvizConfigEditor-linux-0.0.1-x86_64.AppImage
    destination: bin/some-github-binary-artifact
    hash: sha256:32d1f47a90ec01cb554db06ee5cb3c43b497e98c6704f4392d336e14101c8651
    depends: []
" >> repository.yml

# List the packages we want
echo "webviz-core-components:
  0.5.6
some_github_binary_artifact:
  0.0.1a1" > wcc.yml

# Fill in dependencies
komodo-extract-dep-graph wcc.yml 2022.04.00-py36-rhel7.yml repository.yml > wcc-release.yml

# Lint it
komodo-lint wcc-release.yml repository.yml

mkdir build; cd build

# build
# note that you'll need gnu coreutils, macos coreutils will not work
kmd ../wcc-release.yml                   \
    ../repository.yml                    \
    --download                           \
    --build                              \
    --install                            \
    --jobs 4                             \
    --release wcc-release                \
    --tmp tmp                            \
    --cache cache                        \
    --prefix $(pwd)                      \
    --virtualenv $(which virtualenv)     \
    --locations-config ../locations.yml
```

The binary appears where expected! 🆗

**Comment by anders-kiaer** [comment]
Great - thanks for the recipe. 👍

**Pull request merged**